lib/hanami/mailer.rb

```diff
@@ -38,300 +25,265 @@ def initialize
       txt:  'text/plain'
     }.freeze
 
-    include Utils::ClassAttribute
+    private_constant(:CONTENT_TYPES)
 
+    # Base error for Hanami::Mailer
+    #
     # @since 0.1.0
-    # @api private
-    class_attribute :configuration
-    self.configuration = Configuration.new
+    class Error < ::StandardError
+    end
 
-    # Configure the framework.
-    # It yields the given block in the context of the configuration
+    # Unknown mailer
     #
-    # @param blk [Proc] the configuration block
+    # This error is raised at the runtime when trying to deliver a mail message,
+    # by using a configuration that it wasn't finalized yet.
     #
-    # @since 0.1.0
+    # @since next
+    # @api unstable
     #
-    # @see Hanami::Mailer::Configuration
+    # @see Hanami::Mailer.finalize
+    class UnknownMailerError < Error
+      # @param mailer [Hanami::Mailer] a mailer
+      #
+      # @since next
+      # @api unstable
+      def initialize(mailer)
+        super("Unknown mailer: #{mailer.inspect}. Please finalize the configuration before to use it.")
+      end
+    end
+
+    # Missing delivery data error
     #
-    # @example
-    #   require 'hanami/mailer'
+    # It's raised when a mailer doesn't specify `from` or `to`.
     #
-    #   Hanami::Mailer.configure do
-    #     root '/path/to/root'
-    #   end
-    def self.configure(&blk)
-      configuration.instance_eval(&blk)
-      self
+    # @since 0.1.0
+    class MissingDeliveryDataError < Error
+      def initialize
+        super("Missing delivery data, please check 'from', or 'to'")
+      end
     end
 
+    # @since next
+    # @api unstable
+    @_subclasses = Concurrent::Array.new
+
     # Override Ruby's hook for modules.
-    # It includes basic Hanami::Mailer modules to the given Class.
+    # It includes basic `Hanami::Mailer` modules to the given Class.
     # It sets a copy of the framework configuration
     #
     # @param base [Class] the target mailer
     #
-    # @since 0.1.0
-    # @api private
-    #
-    # @see http://www.ruby-doc.org/core/Module.html#method-i-included
-    def self.included(base)
-      conf = configuration
-      conf.add_mailer(base)
-
-      base.class_eval do
-        extend Dsl
-        extend ClassMethods
-
-        include Utils::ClassAttribute
-        class_attribute :configuration
-
-        self.configuration = conf.duplicate
-      end
-
-      conf.copy!(base)
+    # @since next
+    # @api unstable
+    def self.inherited(base)
+      @_subclasses.push(base)
+      base.extend Dsl
```

Comment:
@mereghost First of all, let me explain the idea of transforming `Hanami::Mailer` from a module to a superclass.

This new version of mailer imposes really strict rules: immutability and Single Responsibility Principle (SRP). Indeed, the only public method is `#deliver` (aliased as `#call`).

Because we want that an object that uses `Hanami::Mailer` must be only a mailer, the inheritance helps us to enforce this rule. When you have a module, you can always include it somewhere, but that breaks SRP.

---

I'm not sure what's your idea, but here's how's designed the actual code.

If you use:

```ruby
[1] pry(main)> class Mailer
[1] pry(main)*   def self.inherited(mailer)
[1] pry(main)*     mailer.class_eval do
[1] pry(main)*       @template = nil
[1] pry(main)*     end
[1] pry(main)*   end
[1] pry(main)*   def self.template(value = nil)
[1] pry(main)*     if value.nil?
[1] pry(main)*       @template
[1] pry(main)*     else
[1] pry(main)*       @template = value
[1] pry(main)*     end
[1] pry(main)*   end
[1] pry(main)* end
[2] pry(main)> class WelcomeMailer < Mailer
[2] pry(main)*   template 'welcome'
[2] pry(main)* end
"welcome"
[3] pry(main)> WelcomeMailer.template # OK
"welcome"
[4] pry(main)> Mailer.template # OK, but I don't like it
nil
```

It doesn't makes sense for `Hanami::Mailer` to expose `.template`, but only for its subclasses:

```ruby
[1] pry(main)> class Mailer
[1] pry(main)*   module Dsl
[1] pry(main)*     def self.extended(mailer)
[1] pry(main)*       mailer.class_eval do
[1] pry(main)*         @template = nil
[1] pry(main)*       end
[1] pry(main)*     end
[1] pry(main)*     def template(value = nil)
[1] pry(main)*       if value.nil?
[1] pry(main)*         @template
[1] pry(main)*       else
[1] pry(main)*         @template = value
[1] pry(main)*       end
[1] pry(main)*     end
[1] pry(main)*   end
[1] pry(main)*   def self.inherited(mailer)
[1] pry(main)*     mailer.extend(Dsl)
[1] pry(main)*   end
[1] pry(main)* end
:inherited
[2] pry(main)> class WelcomeMailer < Mailer
[2] pry(main)*   template 'welcome'
[2] pry(main)* end
"welcome"
[3] pry(main)> WelcomeMailer.template # OK
"welcome"
[4] pry(main)> Mailer.template # OK, because we don't want `Mailer.template`
NoMethodError: undefined method `template' for Mailer:Class
from (pry):24:in `__pry__'
```

Is it clear now?